dali/operators/input/video_input.cc

```diff
@@ -0,0 +1,165 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/operators/input/video_input.h"
+#include <memory>
+#include <vector>
+
+namespace dali {
+
+/// By definition, the input batch size of this Operator is always 1.
+constexpr int input_batch_size = 1;
+
+
+template<>
+bool VideoInput<CPUBackend>::SetupImpl(std::vector<OutputDesc> &output_desc,
+                                       const Workspace &ws) {
+  if (!valid_) {
+    InputOperator<CPUBackend>::HandleDataAvailability();
+    TensorList<CPUBackend> encoded_videos;
+    encoded_videos.set_pinned(device_id_ != CPU_ONLY_DEVICE_ID);
+    frames_decoders_.resize(input_batch_size);
+    auto &thread_pool = ws.GetThreadPool();
+    this->ForwardCurrentData(encoded_videos, thread_pool);
+
+    // Creating FramesDecoders
+    auto sample = encoded_videos[0];
+    auto data = reinterpret_cast<const char *>(sample.data<uint8_t>());
+    size_t size = sample.shape().num_elements();
+    frames_decoders_[0] = std::make_unique<FramesDecoder>(data, size, false);
+
+    assert(output_descs_.empty());
+    DetermineOutputDescs(static_cast<int>(frames_decoders_[0]->NumFrames()));
+
+    // This has to be done for every video file, since we need to know the shape of the frames.
+    if (last_sequence_policy_ == "pad") {
+      InitializePadValue(0);
+    }
+
+    valid_ = true;
+  }
+  output_desc.resize(1);
+  output_desc[0] = output_descs_.front();
+  output_descs_.pop_front();
+  return true;
+}
+
+
+template<>
+void VideoInput<CPUBackend>::RunImpl(Workspace &ws) {
+  auto &output = ws.Output<CPUBackend>(0);
+  bool full_sequence;
+  for (int64_t s = 0; s < output.num_samples(); s++) {
+    auto pad_value =
+            last_sequence_policy_ == "pad" ? std::optional<SampleView<CPUBackend>>(GetPadFrame())
+                                           : std::nullopt;
+    full_sequence = DecodeFrames(output[s], 0, frames_per_sequence_, pad_value);
+    if (!full_sequence) {
+      break;
+    }
+  }
+  if (!full_sequence || frames_decoders_[0]->NextFrameIdx() == -1) {
+    Invalidate();
+  }
+}
+
+
+DALI_SCHEMA(experimental__inputs__Video)
+                .DocStr(
+                        R"code(
+Streams and decodes a video from a memory buffer. To be used with long and high resolution videos.
+
+Returns a batch of sequences of frames, with the layout: ``(N, F, H, W, C)``, where:
+
+* ``N`` - number of sequences in a batch. Currently ``N=max_batch_size``,
+* ``F`` - number of frames in a sequence,
+* ``H`` - height of the frame,
+* ``W`` - width of the frame,
+* ``C`` - number of channels in the frame.
+
+When using ``fn.inputs.video`` operator inside the DALI Pipeline, the user needs to provide the data
+using :meth:`Pipeline.feed_input`. When the Operator is fed with data, the Pipeline can be ran
```

Comment:
```suggestion
using :meth:`Pipeline.feed_input`. When the Operator is fed with data, the Pipeline can be run
```

Reply:
Shouldn't it be a past form here? :)

Like

> "When the Operator is fed with data and the Pipeline can be..."

> "...assigned as peculiar"
> "...multiplied by 42"
> "...covered with snow"

Reply:
Yep, but it `can be done` not `can be did`. :D

Reply:
Yeah, it's run :)

Reply:
Oh, that's right ;)